### PR TITLE
React/dp 10134  refactor backstop scenarios

### PR DESCRIPTION
**Pull request by pmallett-mc**
## Description
Updates the mf/react Backstop implementation to discover all stories by parsing each story module to find the `kind` and `storyName` values instead of assuming they will match the path and module dir name.

Additional changes/improvements:
* added better documentation in /mayflower/react/docs/visual-regression.md
* added backstop:approve npm run command to facilitate update workflow (doc updated to use npm run commands mostly instead of docker-compose directly)
* incorporated Steven's suggestion to use path.resolve for backstop paths
* added support for a --debug flag to help with local debugging

## Related Issue / Ticket
- [JIRA issue https://jira.mass.gov/browse/DP-10134](https://jira.mass.gov/browse/DP-10134)

## Steps to Test
Tests in CI should pass for this branch -- other pending PRs will need to update their reference images after this is merged.

If you would like to test locally:
1. cd react
1. git checkout react/DP-10134--refactor-backstop-scenarios
1. npm install
1. npm run backstop # all tests should pass, run takes ~3min
1. edit a component to test for a failure. eg. remove the className prop in Button
1. npm run backstop # the Button atom test should fail
1. open backstop/data/html_report/index.html

## Screenshots
N/A

## Additional Notes:
N/A

#### Impacted Areas in Application

* react backstop tests

#### @TODO

*

### CODE REVIEW

**Comment by pmallett-mc** [comment]
@smurrayatwork pointed out that we may be able to discover kind & storyNames using the Storybook API: https://storybook.js.org/addons/api/
But I am not seeing any way to get that into the node script to create the scenarios.
I found this issue asking for a way to do it with no resolution other than to "do what Storyshots does"
https://github.com/storybooks/storybook/issues/1208

Unless Im missing something, it looks like Storyshots does the naive thing I was doing before: assume the filename is the only test story name: https://github.com/storybooks/storybook/blob/master/addons/storyshots/storyshots-core/src/Stories2SnapsConverter.js

**Review by smurrayatwork — APPROVED** [review]

**Review by mark-casias — APPROVED** [review]
tested locally and passed.

**Comment by clairesunstudio** [comment]
I don't see atoms like buttons, forms, headings in the screenshots --> https://3854-137521197-gh.circle-artifacts.com/0/backstop/data/html_report/index.html

**Comment by smurrayatwork** [comment]
Changes:
- Updated backstop docker image to version 3.5.10, which supports scenario level viewport overrides.
- You no longer need the backstop atoms command; this will now process all components with component level viewport overrides as needed.
- All references have been updated so that all components are pulled into the backstop command without `--atoms`.

File: react/backstop/backstop.js [hunk 1]
```diff
@@ -3,11 +3,14 @@ const path = require('path');
...
-const processAtoms = (process.argv.indexOf('--atoms') > -1);
 const debug = (process.argv.indexOf('--debug') > -1);
 
 // The value of viewports will depend on whether the --atoms switch was set.
-let viewports;
+const viewports = [
+  { label: 'small_atom', width: 400, height: 250 },
+  { label: 'phone', width: 320, height: 480 },
+  { label: 'tablet', width: 1024, height: 768 }
+];
```
Comment (mrjmd) [inline react/backstop/backstop.js:13]: Should this be an exported const so we're not duplicating these values between ``backstop.js`` and ``storyBookBackstop.js`` ?

File: react/backstop/storyBookBackstop.js [hunk 2]
```diff
@@ -91,20 +99,32 @@ const mapComponents = (components, debug) => components.map((component) => {
...
+    viewports
   );
 });
 
+const isAtom = (component) => {
+  const { filePath } = component;
+  return(filePath.indexOf('/atoms/') > -1)
+    && (filePath.indexOf('table') === -1)
+    && (path.basename(filePath) !== 'Image')
+    && (filePath.indexOf('icons') === -1);
```
Comment (mrjmd) [inline react/backstop/storyBookBackstop.js]: These exceptions could probably use comments explaining why they are needed, as they had in the previous version.

File: react/backstop/storyBookBackstop.js [hunk 3]
```diff
@@ -91,20 +99,32 @@ const mapComponents = (components, debug) => components.map((component) => {
...
+};
+
+
```
Comment (mrjmd) [inline react/backstop/storyBookBackstop.js]: Double white space.

File: react/docs/visual-regression.md [hunk 4]
```diff
@@ -7,34 +7,49 @@ This repository uses [BackstopJS](https://garris.github.io/BackstopJS/) for visu
...
 
 Backstop also runs in CircleCI, so your build may fail if you do not update the reference screenshots when you push changes to github.
+
+### Running tests in isolation
+To isolate one of the suites for verification or troubleshooting, you can run them individually:
+
+Note: any time you change mayflower/react, you must re-run build-storybook or backstop could be making screenshots from an outdated build.
+
+```bash
+npm run backstop:atoms
```
Comment (mrjmd) [inline react/docs/visual-regression.md]: Does this need to be updated?

**Review by mrjmd — CHANGES_REQUESTED** [review]
A couple of small considerations noted in comments, I think updating the docs is the only approval blocker.

**Review by mrjmd — APPROVED** [review]
Changes look good!

**Review by avrilpearl — APPROVED** [review]
worked as described!

File: react/backstop/backstop.js (hunk 1 above)
Comment (pmallett-mc) [reply react/backstop/backstop.js:13]: Yea, I think that would be a good change; move into storyBookBackstop & ref from there instead.

**Pull request merged**